### DISCUSSION TITLE
What is the recommended way to run this using Nix flakes?

### DISCUSSION THREAD

**Discussion by Ashvith10** [FerretDB/FerretDB#3634] [General]
Right now, I am using this:

```nix
shellHook = ''
  ferretdb
```;
```

But with this, I am no longer able to access the shell, which is a minor inconvenience. How do I make sure that the binary is executed in the background, and also that the app exits, once I exit from a nix shell (regardless of it being pure or impure)

**Comment by noisersup** [comment]
Not sure if I understand your goal, but running FerretDB in the background with `&` respects the SIGHUP signal
that is sent when the shell is closed. Redirecting stdin and stder to /dev/null would also make shell much cleaner :)
```
shellHook = ''
  ferretdb > /dev/null 2>&1 &

  # Print pid to check if it's running with `ps -p <pid>` later
  PID=$!
  echo $PID
'';
```
If it somehow doesn't work (though it works in my environment), you could probably achieve the goal by using `trap` at the end of the shellHook:

```
trap "kill -9 $PID" EXIT
```

Please let me know if I didn't understand your goal, or you still encounter any issues.

**Reply by Ashvith10** [reply]
@noisersup sorry for the late response, but what do you think of this:
```nix
{
    description = "Simple URL shortener";
    inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    outputs = inputs@{ flake-parts, ... }:
    flake-parts.lib.mkFlake { inherit inputs; } {
        systems = [
            "x86_64-linux" "x86_64-darwin"
            "aarch64-linux" "aarch64-darwin"
        ];

        perSystem = { config, self', inputs', pkgs, system, ... }: {
            devShells.default = with pkgs; mkShell {
                buildInputs = ([
                    nodejs_20
                    ferretdb
                    postgresql
                ]) ++ (with nodePackages; [ pnpm ]);

                shellHook = ''
                export PGHOST=$PWD/.postgres
                export PGDATA=$PGHOST/data
                export PGDATABASE=postgres
                export PGLOG=$PGHOST/postgres.log
                export LANG="en_US.UTF-8"

                mkdir -p $PGHOST

                if [ ! -d $PGDATA ]
                then
                initdb -U $USER -W
                fi

                if ! pg_ctl status
                then
                pg_ctl start -l $PGLOG -o "--unix_socket_directories='$PGHOST'"
                fi

                if ! psql -U $USER --list | grep -q ferretdb
                then
                createdb -U $USER ferretdb
                fi

                ferretdb&
                trap "pg_ctl stop && pkill ferretdb" EXIT
                '';
            };
        };
    };
}
```

The only issue right now is that I have to deal with the logs messing up `tmux`.

**Reply by noisersup** [reply]
I think `ferretdb > /dev/null 2>&1 &` is what you're missing if you want logs to be completely silenced.

Overall, it seems like your setup works, though not sure what's your goal with it. If you want to run FerretDB on NixOS system then I strongly recommend [FerretDB service module](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=services.ferretdb). If you want to involve in development, I suggest to run our docker images by using tasks from our `Taskfile.yml`. https://github.com/FerretDB/FerretDB/blob/main/CONTRIBUTING.md#making-a-working-copy

**Reply by Ashvith10** [reply]
@noisersup hello, thank you for your reply. My intention is to run FerretDB alongside PostgreSQL in an isolated environment for development. I could try using those NixOS options inside the Flake file.
